src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -19,18 +20,32 @@ namespace opossum {
 std::string ConstantCalculationRule::name() const { return "Constant Calculation Rule"; }
 
 bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  auto column_reference_to_value_map = std::map<LQPColumnReference, AllTypeVariant>();
+  _calculate_expressions_in_tree(node, column_reference_to_value_map);
+  auto tree_changed = _replace_column_references_in_tree(node, column_reference_to_value_map);
```

Comment:
Can it happen that the result from l25 differs from the result of l26? Intuitively I'd say no, so it would make the code clearer to leave out the `tree_changed |=` bit.
Also, we're not talking about "trees" anymore, but about lqp/plans, since the lqp is effectively a graph, not a tree

Reply:
In theory yes, it can happen. Depends a little bit on the outcome of my point below though. If we do remove the unneeded columns from ProjectionNodes after replacing the references (as before), then it can happen here for example:
```SELECT b FROM (SELECT 1 + 5, a, b FROM table_a) as foo```
No replacements were made in l25 (`tree_changed` is false), but in l26 the expression `1 + 5` will get removed from the ProjectionNode (`tree_changed` is true).

Oh and yes, updated tree->lqp, thanks!